ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
     private BlobProperties serverBlobProperties;
     private short lifeVersion;
 
+    private AmbryCache blobMetadataCache;
+    private AmbryCacheEntryStatus blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheNotLookedUp;
+
     /**
      * Construct a FirstGetChunk and initialize it with the {@link BlobId} of the overall operation.
+     * @param blobMetadataCache A cache to save blob metadata for composite blobs
      */
-    FirstGetChunk() {
+    FirstGetChunk(AmbryCache blobMetadataCache) {
       super(-1, new CompositeBlobInfo.ChunkMetadata(blobId, 0L, -1L));
+      this.blobMetadataCache = blobMetadataCache;
+      if (shouldLookupMetadataCache()) {
+        BlobMetadata blobMetadata = (BlobMetadata) blobMetadataCache.getObject(blobId.toString());
+        if (blobMetadata != null) {
+          blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheHit;
+          blobType = BlobType.MetadataBlob;
+          compositeBlobInfo = blobMetadata.getCompositeBlobInfo();
+          blobInfo = blobMetadata.getBlobInfo();
+          lifeVersion = blobInfo.getLifeVersion();
+          serverBlobProperties = blobInfo.getBlobProperties();
+          chunkIndexToBuf = new ConcurrentHashMap<>();
+          chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+          successfullyDeserialized = true;
```

Comment:
hmm, this is tricky, we need to set FirstGetChunk to ChunkState.Complete so we don't have to send requests to server nodes. But that being said, we still have to get chunkMetadataList and set chunkIdIterator.  Check out code from line 1616 to line 1636 and method initializeDataChunks.

Reply:
Hmm, are you going with approach of checking cache in GetBlobOperation (and modifying first request to fetch blobInfo instead of metadata chunk) instead of checking cache in NonBlockingRouter to issue separate GetBlobInfo requests? Still thinking that if we go with latter, we wouldn't need with modifications in other methods such as `getOperationFlag`, `handleBody`, etc :)

Reply:
To summarise for others:
1/ Check cache in GetBlobOperation and modify code related to FirstGetChunk
2/ Check cache at the Router level. If cache hit, issue GetBlobInfo followed by GetBlob. If cache miss, just issue GetBlob.
Yes I am going with the first approach. All core changes are within FirstGetChunk class. Modifications to methods getOperationFlag, handleBody aren't too difficult. Second approach may avoid this but may require changes elsewhere. We would just be shifting the changes to another place and also scatter it across classes probably. Also, second approach requires callbacks within callbacks which is confusing.